client/boot/common.js

```diff
@@ -36,41 +36,34 @@ const switchUserLocale = ( currentUser, reduxStore ) => {
 
 const setupContextMiddleware = reduxStore => {
 	page( '*', ( context, next ) => {
-		const parsed = url.parse( location.href, true );
-
-		// Decode the pathname by default (now disabled in page.js)
-		context.pathname = decodeURIComponent( context.pathname );
-
-		context.store = reduxStore;
-
-		// Break routing and do full load for logout link in /me
-		if ( context.pathname === '/wp-login.php' ) {
-			window.location.href = context.path;
-			return;
-		}
-
-		// set `context.query`
-		const querystringStart = context.canonicalPath.indexOf( '?' );
-
-		if ( querystringStart !== -1 ) {
-			context.query = qs.parse( context.canonicalPath.substring( querystringStart + 1 ) );
-		} else {
-			context.query = {};
-		}
-
+		// page.js url parsing is broken so we had to disable it with `decodeURLComponents: false`
+		const parsed = url.parse( context.canonicalPath, true );
+		context.pathname = parsed.pathname;
 		context.prevPath = parsed.path === context.path ? false : parsed.path;
+		context.path = parsed.path;
+		context.query = parsed.query;
 
+		context.hashstring = ( parsed.hash && parsed.hash.substring( 1 ) ) || '';
 		// set `context.hash` (we have to parse manually)
-		if ( parsed.hash && parsed.hash.length > 1 ) {
+		if ( context.hashstring ) {
 			try {
-				context.hash = qs.parse( parsed.hash.substring( 1 ) );
+				context.hash = qs.parse( context.hashstring );
 			} catch ( e ) {
 				debug( 'failed to query-string parse `location.hash`', e );
 				context.hash = {};
 			}
 		} else {
 			context.hash = {};
 		}
+
+		context.store = reduxStore;
+
+		// Break routing and do full load for logout link in /me
+		if ( context.pathname === '/wp-login.php' ) {
```

Comment:
I'm not sure I understand why do we need a special handling of that here,
since we [have it here](https://github.com/Automattic/wp-calypso/blob/master/client/me/sidebar/index.jsx#L49-L71).

Could you elaborate?

Reply:
I'm not sure, it seems to be related to `config( 'logout_url' )` we're supposed to redirect the user there on logout but maybe in some case we land back on calypso. I just moved that part here so I'm not sure.